### ISSUE TITLE
Random comments

### ISSUE THREAD

**Issue by ilevkivskyi** [ericvsmith/dataclasses#9]
Sorry, it is not possible to comment directly in code, here are some random comments (particularly for TODO items):

1)
```
#  what exception to raise when non-default follows default? currently
#  ValueError
```
typing module raises ``TypeError``, we might want to synchronize this.

2)
```
#  what to do if a user specifies a function we're going to overwrite,
#  like __init__? error? overwrite it?
```
typing module defines some predefined set of "infrastructure-critical" attributes,
It raises ``AttributeError("Cannot overwrite NamedTuple attribute " + key)`` for these,
and allows overwriting others. I think we should do the same here.

3)
```
#  use typing.get_type_hints() instead of accessing __annotations__
#  directly? recommended by PEP 526, but that's importing a lot just
#  to get at __annotations__
```
I think it is absolutely OK to directly use ``__annotations__`` here as far as
we don't do anything fancy/dangerous with ``__annotations__``. These classes
should be as fast as possible.

4)
```
# is __annotations__ guaranteed to be an ordered mapping?
```
PEP 526 allows this (emphasis my):
> ``__annotations__`` attribute of that module or class (mangled if private) as an _ordered_ mapping from names to evaluated annotations.

5)
> Support setting ``__slots__``.

I am thinking maybe ``__slots__`` should be added by default, with an option to  not add them?

6)
> Optional support by type checkers (mypy) and typing.py (in the sense that you can declare and use data classes without using typing).

mypy supports/understands descriptor protocol, so that a custom stub can be added to typeshed that will make mypy understand dataclasses semantics correctly (also some special-casing in mypy can be added for even better support, I can take care of this when we decide on the semantics/implementation).

7)
> Provide a way to detect if an object is an instance of any data class, similar to using _fields on namedtuples. I'm not sure why people want to be able to detect this, but they do, and Raymond has always suggested looking for an attribute "_fields".

People might want some structural subtyping support for dataclasses. Potentially we can support ``hasattr`` or/and a custom ``__subclasshook__`` in spirit of ``collections.abc`` and PEP 544. Static support for this is also possible (again in spirit of PEP 544, I hope mypy will support it soon, see https://github.com/python/mypy/pull/3132).

**EDIT:** enumerated comments to simplify reading.

**Comment by ericvsmith** [comment]
Thanks, @ilevkivskyi. These are very helpful.

1: I'll change it to TypeError to match typing.
2: I think that's a good idea. I'll open an issue.
3: Okay, I'll remove the TODO item.
4: Thanks. I 'm not sure why I couldn't find that sentence, despite looking for it. I've removed the comments in the code.
5: I've opened issue #10.
6: My main point here was to not require typing.py, but it's good to know it will work. Since the decorator always returns the same class as its argument (or, in the case of `__slots__`, an equivalent class), it shouldn't be much work.
7: I'm adding a `__dataclass_fields__`, which I use to detect which base classes have fields. Also, I'll add something similar to attrs' `attr.fields()` and/or `attr.has()`, which will need to use this. So I think the question of "does there need to be a marker" is yes, and we can argue about the exact names.
